### ISSUE TITLE
[googletts] Only white noise

### ISSUE THREAD

**Issue by DJPsycho82** [openhab/openhab-addons#10015]
Platform information:
Hardware: Raspberry Pi 4B 2GB
OS: Openhabian 1.6.2 64Bit
Java Runtime Environment: Java 11
openHAB version: 3.0.1 release build

When openhab service is stopped the voice played bij Google TTS service via 3.5mm jack output, i only hear white noise. Sometimes rebooting the raspberry pi solves the problem. Every time I restart the openhab service, this problem occurs again.

What I checked:
- All google voices are present
- I can select them in the UI, i tried a couple Wavenet and the ‘default-ones’, doesn’t change
- VoiceRSS works fine
- The wav files (temp folder) play back nicely when opened on another device
- Google limits is ok
- I’m just using openhab:voice say test on karaf > no difference
- openhab:audio play barking.mp3 < works fine
- playback via chromecast is fine

Not sure how to fix it.

***-  At this very moment, I had to reboot the Pi > White noise >restarted openhab > white noise when 'saying' something... I cleared the cache with [openhab-cli clean-cache] >>> This solved the problem this time

**Comment by DJPsycho82** [comment]
Just a bump. Can I help in any kind of way to sort this problem. It really is a problem for me, so I want to help any way i can. But I don't know that to do anymore. :)

**Comment by cweitkamp** [comment]
The add-on stores the sound files generated by Googles TTS in a local folder. You can find them in the `$OPENHAB_USERDATA/cache/org.openhab.voice.googletts`. Did you purge them over there too? If not please remove the folder and try it again.

**Comment by DJPsycho82** [comment]
Yes, when I clear cache the folder is empty afterwards but still white noise. Strange thing now is. I can't get it to work anymore. I rebooted, clean cache, restarted openhab service, stopped the bundle via karaf, stopped all other voice related bundles, uninstalled it via add-ons.cfg. but it just gives noise, but only via the 3mm port, via Chromecast the voice works just normal. The sound files in the cache play just fine, so the whole "get Voice from Google"-part works. Just the playback via 3mm Jack gives noise. And only Google TTS. Other mp3's just work fine when Google TTs fails.

**Comment by cweitkamp** [comment]
Are you using "System Speaker" as a default audio sink? Do you have other audio sinks for testing available? Does it work with them?

**Comment by DJPsycho82** [comment]
Yes, I meant "System speaker" when I said 3mm jack. I tested it with Chromecast audio sink and then it just works. No more noise, but when I change to System speaker i only hear noise.

**Comment by cweitkamp** [comment]
I have the slight feeling it is not related to Google TTS. But javasound. Can you enable DEBUG logging for `org.openhab.voice.googletts` and `org.openhab.core.audio` while trying to "say" something and provide the logs over here?

**Comment by DJPsycho82** [comment]
I think you're right. I didn't know javasound could be the problem, so I only was able to focus on Google TTS.

```
==> /var/log/openhab/openhab.log <==

2021-03-12 18:32:37.388 [DEBUG] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Synthesize 'h ' for voice 'googletts:nlNLWavenetA' in format AudioFormat [codec=PCM_SIGNED, container=WAVE, bigEndian=true, bitDepth=16, bitRate=705600, frequency=44100]

2021-03-12 18:32:37.390 [DEBUG] [org.openhab.voice.googletts.internal.GoogleCloudAPI     ] - Audio file nlNLWavenetA_e706061669e0a3ec967386145b286b47.wav was found in cache.

2021-03-12 18:32:37.410 [WARN ] [openhab.core.audio.internal.javasound.JavaSoundAudioSink] - Cannot determine master volume level - assuming 100%
```

**Comment by DJPsycho82** [comment]
Hmm, when i check the file it outputs it has these  properties: 16bit,1 channel, 24000hz. But that's not what the log says.

**Comment by cweitkamp** [comment]
I think we got a track. My log is different - TTS service uses a different audio format (mp3):

```
2021-03-12 19:51:48.738 [DEBUG] [.googletts.internal.GoogleTTSService] - Synthesize 'Hoe is het met je? ' for voice 'googletts:nlNLWavenetA' in format AudioFormat [codec=MP3, container=NONE, bitDepth=16, bitRate=64000, frequency=44100]
2021-03-12 19:51:49.085 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Caching audio file nlNLWavenetA_d5764ad47f5d391466b832c4ea159c5e.mp3
2021-03-12 19:51:49.087 [DEBUG] [ce.googletts.internal.GoogleCloudAPI] - Caching text file nlNLWavenetA_d5764ad47f5d391466b832c4ea159c5e.txt
2021-03-12 19:51:49.106 [WARN ] [nternal.javasound.JavaSoundAudioSink] - Cannot determine master volume level - assuming 100%
```

When increasing logging to TRACE for googletts and restarting the bundle do you see something like this:

```
2021-03-12 20:05:34.957 [TRACE] [.googletts.internal.GoogleTTSService] - Initializing audio formats
2021-03-12 20:05:34.958 [WARN ] [.googletts.internal.GoogleTTSService] - Audio format OGG_OPUS is not yet supported.
2021-03-12 20:05:34.958 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format not supported: OGG_OPUS
2021-03-12 20:05:34.959 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format supported: MP3
2021-03-12 20:05:34.960 [TRACE] [.googletts.internal.GoogleTTSService] - Audio format supported: LINEAR16
```

**Comment by DJPsycho82** [comment]
I'm not at home right now, but I'm sure I have this in my logs every time Google TTS is used:

format not supported: OGG_OPUS

For the other TRACE logs I have to check my system. Will do this evening. Looks like the system(javasound I guess) is playing the file in a wrong format. Funny thing is, you can hear the words kind of when noise is being played.

(By the way, I'm also from Holland, and (het gaat goed) I'm doing fine haha.

**Comment by DJPsycho82** [comment]
Here is my output with TRACE, no strange extra's when using voice (noise lol):

```
2021-03-12 21:24:47.365 [DEBUG] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Synthesize 'avondklok, nu weer thuis ' for voice 'googletts:nlNLWavenetA' in format AudioFormat [codec=PCM_SIGNED, container=WAVE, bigEndian=true, bitDepth=16, bitRate=705600, frequency=44100]

2021-03-12 21:24:47.620 [DEBUG] [org.openhab.voice.googletts.internal.GoogleCloudAPI     ] - Caching audio file nlNLWavenetA_8e78b61c0c80806749139c532a96be5d.wav

2021-03-12 21:24:47.621 [DEBUG] [org.openhab.voice.googletts.internal.GoogleCloudAPI     ] - Caching text file nlNLWavenetA_8e78b61c0c80806749139c532a96be5d.txt

2021-03-12 21:24:47.636 [WARN ] [openhab.core.audio.internal.javasound.JavaSoundAudioSink] - Cannot determine master volume level - assuming 100%
```

And below, same output as yours.

```
2021-03-12 21:26:01.944 [DEBUG] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Using cache folder /var/lib/openhab/cache/org.openhab.voice.googletts

2021-03-12 21:26:01.949 [DEBUG] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Updating configuration

2021-03-12 21:26:01.950 [TRACE] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - New configuration: GoogleTTSConfig{pitch=0.0, speakingRate=1.0, volumeGainDb=0.0, purgeCache=false}

2021-03-12 21:26:02.042 [TRACE] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Initializing voices

2021-03-12 21:26:02.045 [TRACE] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Google Cloud TTS voice: pl-PL-Standard-E
.
.
.
2021-03-12 21:26:02.292 [TRACE] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Initializing audio formats

2021-03-12 21:26:02.293 [WARN ] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Audio format OGG_OPUS is not yet supported.

2021-03-12 21:26:02.294 [TRACE] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Audio format not supported: OGG_OPUS

2021-03-12 21:26:02.295 [TRACE] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Audio format supported: MP3

2021-03-12 21:26:02.296 [TRACE] [org.openhab.voice.googletts.internal.GoogleTTSService   ] - Audio format supported: LINEAR16
```

Get's me thinking, could there be anywhere a config file where the parameters could be set? Maybe there is somewhere a place where the output settings can change. Because the mp3's file frequency is 24000hz, but the log says: [codec=PCM_SIGNED, container=WAVE, bigEndian=true, bitDepth=16, bitRate=705600, frequency=44100]. Not sure if that's cosmetical or just wrong.

**Comment by DJPsycho82** [comment]
Some extra info:

		<Logger level="INFO" name="org.openhab.voice.googletts"/>
		<Logger level="TRACE" name="org.openhab.core.audio"/>
		<Logger level="OFF" name="core.audio.internal.AudioManagerImpl"/>
		<Logger level="OFF" name="openhab.core.audio.internal.javasound.JavaSoundAudioSink"/>

I set the log levels from above to TRACE via Karaf, but when playing, there are no errors. I don't think the problem is within the GoogleTTS binding. But I don't know where to look to check the Enhanced Javasound for errors or settings...

**Comment by cweitkamp** [comment]
All involved bundles are supporting both audio formats mp3 and wav whereas mp3 is the preferred one. I am not aware of any configuration setting to override the default values. Currently I do not have a clue where to look next. I do not think that it is a hardware problem.

ftr. I am not from the Netherlands. 😉 Ik ben Duits (I am German).

**Comment by DJPsycho82** [comment]
Me neither... Hopefully someone will find a solution for this after reading this thread. If I stumble upon something I will post it here but I'm out of options too.

(Not many Germans speak Dutch. Nice to see our language 'across the borders' :D.)

**Comment by BobMiles** [comment]
I'm pretty sure it's not a hardware problem either, I'm on a rpi3b+.
I would even say the problems for me began without even restarting openhab. Suddenly it was just white noise and ever since I couldn't get it back to work. Alsa & Co seem fine...

**Comment by DJPsycho82** [comment]
With hardware I meant driver for certain Pi model but Nice to hear it isn't only the pi4 but also other models. So it had to do something with playback, but how and where it goes wrong...

**Comment by hanzz** [comment]
Also hit this and second restart fixed it. It seems whenever TTS uses .wav it does not work. If it uses mp3 it does work. I've started digging into code more. This is just dump of my process. I hope in the end I will find out what's wrong :).

### How does google-tts openhab service work?

- [tts.getSupportedFormat](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java#L211) returns two formats for googletts - AudioFormat.CONTAINER_WAVE for PCM (wav) and AudioFormat.CONTAINER_NONE for mp3.
- [getBestMatch](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java#L408) is called [here](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java#L217).
- getBestMatch calls [getPrefferedFormat](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java#L313).
- getPreferredFormat prefers CONTAINER_WAVE (it does not return anything else than WAVE).
- getPreferredFormat at first checks if that CONTAINER_WAVE has *all* the values set. If it does, it returns the format and uses it. If it does not, it sets some values to defaults and returns it. I think that's the root issue here:

GoogleTTSService creates the CONTAINER_WAVE like this:

```
                return new AudioFormat(AudioFormat.CONTAINER_WAVE, AudioFormat.CODEC_PCM_SIGNED, null, bitDepth, null,
                        frequency);
```

The `null` values there are for `bigEndian` and `bitRate`.

- In this case when `bigEndian` is `null`, the getPreferredFormat set it the TRUE [here](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java#L364).
- In this case when `bitRate` is `null`, it also computes bitRate as `bitRate = Integer.valueOf(bitDepth.intValue() * frequency.intValue());`

### What is the .wav produced by it?

I've unfortunately removed the cache, but I have this is my terminal history. I have no reason to not believe `file` for now:

```
# file csCZStandardA_f80e4f8a8040ff6f79f28a81429514b5.wav 
csCZStandardA_f80e4f8a8040ff6f79f28a81429514b5.wav: RIFF (little-endian) data, WAVE audio, Microsoft PCM, 16 bit, mono 24000 Hz
```

### What does google API says for text.synthesize?

https://cloud.google.com/text-to-speech/docs/reference/rest/v1/text/synthesize

LINEAR16 | Uncompressed 16-bit signed *little-endian* samples (Linear PCM). Audio content returned as LINEAR16 also contains a WAV header.

It also seems that sampleRateHertz is by default set according to voice, which for my voice is 24000Hz. This is also what `file` returns for that wav. It is possible to override it, but they say it might not be supported for all possible values. Not sure if setting it to 44100 would work.

### Conclusion

I think there are issues in GoogleTTSService plugin when creating LINEAR16 audioformat in `GoogleTTSService.getAudioFormat`:

1) I think it should set `frequency` according to `naturalSampleRateHertz` of voice (preferrably) or it should try using 44100 in `AudioConfig.sampleRateHertz` or it should examine the resulting `.wav` and construct the format based on that (well, this might not be possible, because you need to create format before requesting the .wav) or it should stop using CONTAINER_WAVE completely and just use MP3. Speaking about MP3, I think the AudioContainer is also not right, but the whathever Openhab uses to play it is probably smart enough when it comes to MP3 to just play it. ffprobe says `Stream #0:0: Audio: mp3, 24000 Hz, mono, fltp, 32 kb/s` it is not what is set for MP3 audioFormat.
2) I think it should set `bigEndian` to FALSE.


### Why does it sometimes work?

My theory is that [getBestMatch](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java#L408) sometimes contains MP3 before WAV in `outputs`. If that happens mp3 is used, if WAV appears first in `output`, then WAV is used and it's broken.

This is quite possible, because MP3/WAV is defined in Set here: https://github.com/openhab/openhab-core/blob/ee1d3f3a73c728973e8bb374641615f42f1dc2f9/bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/internal/javasound/JavaSoundAudioSink.java#L68

Reading some doc about Java sets, it seems `The elements are returned in no particular order (unless this set is an instance of some class that provides a guarantee).`

**Comment by sjcliffe** [comment]
I've just upgraded from 2.5.8 to 3.0.2 and am experiencing this issue on my Raspberry Pi 4. 

The generated WAV file in the cache folder play fine using aplay on the machine.

**Comment by Gransi** [comment]
I have the same issue. After several service restarts google TTS works normal. 

Hardware: AMD Ryzen 3200G System
OS: Ubuntu 20.04
Java Runtime Environment: Java 11
openHAB version: 3.1.0 release build

**Comment by hanzz** [comment]
FYI, I've switched to voicerss openhab addon which works as expected for my use-cases. It supports lot of languages and does not have this issue. With google tts, I also had an issue when I had to add the token after few days to get it work. I suggest people to do the same.

**Comment by DJPsycho82** [comment]
I didn't like voicerss because of the bad sound quality, it sounded as if the speaker was packed in a large blanket, a verry low sample/bit rate. But I checked again today and saw 2 new voices for dutch. I tried one of them and it sounds pretty good. Ok, wavenet from google is just better but this is good enough for me. I still hope someone can fix google TTS one day but for the time being VoiceRSS is fine for me.

**Comment by sjcliffe** [comment]
Voicerss looks and sounds good - I tried it via their website. Unfortunately after installing the binding it seems you can't change from the default voice :-(

**Comment by DJPsycho82** [comment]
It's a bit off topic. But it's right, voicerss is at a high nitrate on their website. But you can change the voice in the settings of openhab. I did it yesterday. For Dutch there are 3 voices, 1 default, 1 female and 1 male voice. Only the male voice sounds clear for Dutch. I'm pretty sure for English there are more then 1 voices to choose from.

**Comment by DJPsycho82** [comment]
I updated my openhab to version 3.2 just an hour ago and the white noise problem still exists. Did not try anything fancy but just booted the new version, changed voice to google TTS and executed openhab:voice say h via Karaf and it gave a white noise.

**Comment by dalgwen** [comment]
Hello all.
I found a bug with the way WAV sound file is played in openHAB.
The issue seems to be here [here](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/FileAudioStream.java#L60)
openHAB suppose a specific format for the WAV file instead of parsing the file and getting the correct one.
After that, when passed to the javasound / system speaker audio sink implementation, the "java player" is set to play this (incorrectly guessed) file format.
I ran into to a sideway issue when I made the pulseaudio audio sink. The [bug report is here](https://github.com/openhab/openhab-addons/issues/11630). In this bug, I needed to know if I had to convert a WAV file to the pulseaudio default format, and as the information provided by openHAB was wrong, I had to implement a parsing method (the [PR is here](https://github.com/openhab/openhab-addons/pull/11631)).

I suppose the parsing method could be extracted and used here.

I'm not totally sure it will resolve the issue (white noise is strange, I would expect an incorrect speed or pitch instead). But it is nonetheless a required step in the right direction.
